### ISSUE TITLE
[BUG] Volume operations/rebuilding error during eviction

### ISSUE THREAD

**Issue by shuo-wu** [longhorn/longhorn#4294]
## Describe the bug

While evicting replicas, the engine process may not remove the deprecating replicas when receiving the operation requests. Then trying to send requests to the deprecating replicas would cause rpc errors.

This issue will lead to [the snapshot purging operation after the replica rebuilding triggered by the eviction itself](https://github.com/longhorn/longhorn-manager/blob/16cfd1c03699f195ed520ef4a8f634d99474c9d4/controller/engine_controller.go#L1551-L1559) erroring out. But actually, the purging is issued successfully to all healthy replicas.

## To Reproduce

1. Create and attach a multi-replica volume
2. Prepare one extra disk for a node that contains at least one volume replica
3. Evicting the old disk for node
4. Check the engine YAML or longhorn manages log

## Expected behavior

No errors in the engine YAML or logs.

## Log or Support bundle

Even if the auto snapshot purge for rebuilding is disabled, we are still able to find errors in the engine YAML
```
apiVersion: longhorn.io/v1beta2
kind: Engine
metadata:
  creationTimestamp: "2022-07-27T04:46:03Z"
  finalizers:
  - longhorn.io
  generation: 4
  labels:
    longhornnode: shuo-k8s-worker-1
    longhornvolume: vol1
  name: vol1-e-998b62c7
  namespace: longhorn-system
  ownerReferences:
  - apiVersion: longhorn.io/v1beta2
    kind: Volume
    name: vol1
    uid: fe656464-78af-4abf-8068-0742ba247fef
  resourceVersion: "34220387"
  uid: f6a339e2-d606-479f-910f-c787f9efa906
spec:
  active: true
  backupVolume: ""
  desireState: running
  disableFrontend: false
  engineImage: longhornio/longhorn-engine:master-head
  frontend: blockdev
  logRequested: false
  nodeID: shuo-k8s-worker-1
  replicaAddressMap:
    vol1-r-769e039f: 10.42.4.30:10000
    vol1-r-56144d78: 10.42.2.81:10000
    vol1-r-8724804e: 10.42.1.59:10000
  requestedBackupRestore: ""
  requestedDataSource: ""
  revisionCounterDisabled: false
  salvageRequested: false
  upgradedReplicaAddressMap: {}
  volumeName: vol1
  volumeSize: "1073741824"
status:
  backupStatus: null
  cloneStatus:
    tcp://10.42.1.59:10000:
      error: ""
      fromReplicaAddress: ""
      isCloning: false
      progress: 0
      snapshotName: ""
      state: ""
    tcp://10.42.2.81:10000:
      error: ""
      fromReplicaAddress: ""
      isCloning: false
      progress: 0
      snapshotName: ""
      state: ""
    tcp://10.42.4.30:10000:
      error: ""
      fromReplicaAddress: ""
      isCloning: false
      progress: 0
      snapshotName: ""
      state: ""
    tcp://10.42.4.30:10015:
      error: 'failed to get snapshot clone status of tcp://10.42.4.30:10015: failed
        to get snapshot clone status: rpc error: code = Unavailable desc = all SubConns
        are in TransientFailure, latest connection error: connection error: desc =
        "transport: Error while dialing dial tcp 10.42.4.30:10017: connect: connection
        refused"'
      fromReplicaAddress: ""
      isCloning: false
      progress: 0
      snapshotName: ""
      state: ""
  currentImage: longhornio/longhorn-engine:master-head
  currentReplicaAddressMap:
    vol1-r-769e039f: 10.42.4.30:10000
    vol1-r-56144d78: 10.42.2.81:10000
    vol1-r-8724804e: 10.42.1.59:10000
  currentSize: "1073741824"
  currentState: running
  endpoint: /dev/longhorn/vol1
  instanceManagerName: instance-manager-e-3bdc3f00
  ip: 10.42.4.31
  isExpanding: false
  lastExpansionError: ""
  lastExpansionFailedAt: ""
  lastRestoredBackup: ""
  logFetched: false
  ownerID: shuo-k8s-worker-1
  port: 10001
  purgeStatus:
    tcp://10.42.1.59:10000:
      error: ""
      isPurging: false
      progress: 0
      state: ""
    tcp://10.42.2.81:10000:
      error: ""
      isPurging: false
      progress: 0
      state: ""
    tcp://10.42.4.30:10000:
      error: ""
      isPurging: false
      progress: 0
      state: ""
    tcp://10.42.4.30:10015:
      error: ""
      isPurging: false
      progress: 0
      state: ""
  rebuildStatus: {}
  replicaModeMap:
    vol1-r-769e039f: RW
    vol1-r-47418d68: RW
    vol1-r-56144d78: RW
    vol1-r-8724804e: RW
  restoreStatus:
    tcp://10.42.1.59:10000:
      backupURL: ""
      currentRestoringBackup: ""
      isRestoring: false
      lastRestored: ""
      state: ""
    tcp://10.42.2.81:10000:
      backupURL: ""
      currentRestoringBackup: ""
      isRestoring: false
      lastRestored: ""
      state: ""
    tcp://10.42.4.30:10000:
      backupURL: ""
      currentRestoringBackup: ""
      isRestoring: false
      lastRestored: ""
      state: ""
    tcp://10.42.4.30:10015:
      backupURL: ""
      currentRestoringBackup: ""
      error: 'Failed to get restoring status on tcp://10.42.4.30:10015: failed to
        get restore status: rpc error: code = Unavailable desc = all SubConns are
        in TransientFailure, latest connection error: connection error: desc = "transport:
        Error while dialing dial tcp 10.42.4.30:10017: connect: connection refused"'
      isRestoring: false
      lastRestored: ""
      state: ""
  salvageExecuted: false
  snapshots:
    35491870-d26d-4083-abf5-8fe36453eaec:
      children:
        volume-head: true
      created: "2022-07-27T04:50:53Z"
      labels: {}
      name: 35491870-d26d-4083-abf5-8fe36453eaec
      parent: ""
      removed: false
      size: "0"
      usercreated: false
    volume-head:
      children: {}
      created: "2022-07-27T04:50:53Z"
      labels: {}
      name: volume-head
      parent: 35491870-d26d-4083-abf5-8fe36453eaec
      removed: false
      size: "0"
      usercreated: false
  snapshotsError: ""
  started: true
  storageIP: 10.42.4.31
```

## Environment

 - Longhorn version: v1.3.0

## Additional context

To fix this issue, there may be 2 options:
1. Waiting for engine processes to remove invalid replicas before handling operations.
2. Ignore the errors triggered by the invalid replicas in the longhorn manager side.

**Comment by shuo-wu** [comment]
There are 2 issues here:
1. As mentioned in the ticket description, there will be some transient error messages in all kinds of engine status maps. These error messages are misleading but they won't affect the functionality.
2. After the rebuilding is complete, the subsequent purge will be issued. But the purge would be failed by the shutdown of the evicting replica:
    1. Currently, when Longhorn deletes or removes a running replica from a running engine, the replica process will be directly removed first. Then the engine process monitors the replica process failure then report the unavailability to the longhorn manager. In the end, the longhorn manager will help remove the replica record from the engine CR. IMO, this kind of deletion workflow looks weird. Thinking about the volume attachment workflow, Longhorn always starts the replica processes then asks the engine to monitor them. The removal should be done in a reverse way: Longhorn manager asks the engine to stop the replica monitoring and remove the record then shut down the replica process.
    2. Even if the above flow is fixed. The purge command may still get failed as since it's not an atomic operation. The engine monitoring stop can happen between [any of these operations](https://github.com/longhorn/longhorn-engine/blob/c0d826a25a73ada50d6beb13ee6b0cd3b5056e2a/pkg/sync/sync.go#L145-L192) then leads to failure. I am thinking if the best fix at this moment is adding retries for the purge invocation.

**Comment by shuo-wu** [comment]
After discussing with @innobead, we plan to fix 1-1 in v1.3.1.
For 2-1, it's better to open a separate ticket to track it. Since it impacts a lot, it won't be resolved in v1.3.1.
As for 2-2, there may be no need to add a retry. Regular snapshot deletion would help purge all system snapshots.

**Comment by yangchiu** [comment]
Verified passed on master-head (longhorn-manager 9746018) by following [the test steps](https://github.com/longhorn/longhorn-tests/pull/1067/files).
